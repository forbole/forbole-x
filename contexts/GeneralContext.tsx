import React from 'react';
import get from 'lodash/get';
import currencies from '../misc/currencies';
import usePersistedState from '../misc/usePersistedState';

type Theme = 'light' | 'dark';
type Currency = typeof currencies[number];
type FavValidators = string[];
type FavAddress = { address: string; crypto: string; moniker: string; note?: string; img?: string };
interface GeneralState {
  currency: Currency;
  currencyRate: number; // per USD rate
  theme: Theme;
  alwaysRequirePassword: boolean;
  setAlwaysRequirePassword?: React.Dispatch<React.SetStateAction<boolean>>;
  hideAmount: boolean;
  setHideAmount?: React.Dispatch<React.SetStateAction<boolean>>;
  favValidators: FavValidators;
  favAddresses: FavAddress[];
  setCurrency?: React.Dispatch<React.SetStateAction<Currency>>;
  setTheme?: React.Dispatch<React.SetStateAction<Theme>>;
  addFavValidators?: (id: string) => void;
  deleteFavValidators?: (id: string) => void;
  addFavAddresses?: (n: FavAddress) => void;
  deleteFavAddresses?: (id: string) => void;
  updateFavAddresses?: (n: FavAddress) => void;
}

const initialState: GeneralState = {
  currency: 'USD',
  currencyRate: 1,
  theme: 'light',
  alwaysRequirePassword: false,
  hideAmount: false,
  favValidators: [],
  favAddresses: [],
};

const GeneralContext = React.createContext<GeneralState>(initialState);

const GeneralProvider: React.FC = ({ children }) => {
  const [currency, setCurrency] = usePersistedState('currency', initialState.currency);
  const [currencyRate, setCurrencyRate] = usePersistedState(
    'currencyRate',
    initialState.currencyRate,
  );
  const [theme, setTheme] = usePersistedState('theme', initialState.theme);
  const [alwaysRequirePassword, setAlwaysRequirePassword] = usePersistedState(
    'alwaysRequirePassword',
    initialState.alwaysRequirePassword,
  );
  const [hideAmount, setHideAmount] = usePersistedState('hideAmount', initialState.hideAmount);

  const [favValidators, setFavValidators] = usePersistedState('fav', initialState.favValidators);
  const [favAddresses, setFavAddresses] = usePersistedState(
    'favAddress',
    initialState.favAddresses,
  );
  const addFavValidators = React.useCallback(
    (address: string) => {
      setFavValidators(vs => [address, ...vs]);
    },
    [setFavValidators],
  );

  const deleteFavValidators = React.useCallback(
    (address: string) => {
      setFavValidators(vs => vs.filter(a => a !== address));
    },
    [setFavValidators],
  );

  const addFavAddresses = React.useCallback(
    (newAddress: FavAddress) => {
      setFavAddresses(vs => [newAddress, ...vs]);
    },
    [setFavAddresses],
  );

  const deleteFavAddresses = React.useCallback(
    (address: string) => {
      setFavAddresses(vs => vs.filter(a => a.address !== address));
    },
    [setFavAddresses],
  );

  const updateFavAddresses = React.useCallback(
    (editedAddress: {
      address: string;
      crypto: string;
      moniker: string;
      note: string;
      img: string;
      newAddress: string;
    }) => {
      setFavAddresses(vs =>
        vs.map(a =>
          a.address === editedAddress.address
            ? {
                address: editedAddress.newAddress,
                crypto: editedAddress.crypto,
                moniker: editedAddress.moniker,
                note: editedAddress.note,
                img: editedAddress.img,
              }
            : a,
        ),
      );
    },
    [setFavAddresses],
  );

  React.useEffect(() => {
    if (currency === 'USD') {
      setCurrencyRate(1);
    } else {
      fetch(
        `${process.env.NEXT_PUBLIC_ALPHAVANTAGE_API_URL}/query?function=CURRENCY_EXCHANGE_RATE&from_currency=USD&to_currency=${currency}&apikey=${process.env.NEXT_PUBLIC_ALPHAVANTAGE_API_KEY}`,
      )
        .then(r => r.json())
        .then(r =>
          setCurrencyRate(
            Number(get(r, ['Realtime Currency Exchange Rate', '5. Exchange Rate'], '')),
          ),
        );
    }
  }, [currency]);

  return (
    <GeneralContext.Provider
      value={{
        currency,
        currencyRate,
        setCurrency,
        theme,
        setTheme,
        alwaysRequirePassword,
        setAlwaysRequirePassword,
        hideAmount,
        setHideAmount,
        favValidators,
        addFavValidators,
        deleteFavValidators,
        favAddresses,
        addFavAddresses,
        deleteFavAddresses,
        updateFavAddresses,
      }}
    >
      {children}
    </GeneralContext.Provider>
  );
};

const useGeneralContext = (): GeneralState => React.useContext(GeneralContext);

export { GeneralProvider, useGeneralContext };
