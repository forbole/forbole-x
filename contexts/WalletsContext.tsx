import React from 'react'
import sendMsgToChromeExt from '../misc/sendMsgToChromeExt'

interface WalletsState {
  isFirstTimeUser: boolean
  isUnlocked: boolean
  wallets: Wallet[]
  accounts: Account[]
  password: string
  unlockWallets?: (password: string) => void
  addWallet?: (wallet: CreateWalletParams) => void
  updateWallet?: (id: string, wallet: UpdateWalletParams) => void
  deleteWallet?: (id: string) => void
  addAccount?: (account: CreateAccountParams, securityPassword: string) => void
  updateAccount?: (address: string, account: UpdateAccountParams) => void
  deleteAccount?: (address: string) => void
  viewMnemonicPhrase?: (id: string, securityPassword: string) => Promise<{ success: boolean }>
  viewMnemonicPhraseBackup?: (
    id: string,
    securityPassword: string,
    backupPassword: string
  ) => Promise<string>
}

const initialState: WalletsState = {
  isFirstTimeUser: false,
  isUnlocked: false,
  wallets: [],
  accounts: [],
  password: '',
}

const WalletsContext = React.createContext<WalletsState>(initialState)

const WalletsProvider: React.FC = ({ children }) => {
  const [wallets, setWallets] = React.useState<Wallet[]>([])
  const [accounts, setAccounts] = React.useState<Account[]>([])
  const [isFirstTimeUser, setIsFirstTimeUser] = React.useState(false)
  const [password, setPassword] = React.useState('')

  const checkIsFirstTimeUser = React.useCallback(async () => {
    const response = await sendMsgToChromeExt({
      event: 'ping',
    })
    setIsFirstTimeUser(response.isFirstTimeUser)
  }, [])

  const unlockWallets = React.useCallback(
    async (pw: string) => {
      if (!isFirstTimeUser) {
        const walletaResponse = await sendMsgToChromeExt({
          event: 'getWallets',
          data: {
            password: pw,
          },
        })
        const accountsResponse = await sendMsgToChromeExt({
          event: 'getAccounts',
          data: {
            password: pw,
          },
        })
        setWallets(walletaResponse.wallets)
        setAccounts(accountsResponse.accounts)
      }
      setPassword(pw)
    },
    [isFirstTimeUser, setPassword, setWallets, setAccounts]
  )

  const addWallet = React.useCallback(
    async (wallet: CreateWalletParams) => {
      const result = await sendMsgToChromeExt({
        event: 'addWallet',
        data: {
          wallet,
          password,
        },
      })
      setIsFirstTimeUser(false)
      setWallets((ws) => [result.wallet, ...ws])
      setAccounts((acs) => [...result.accounts, ...acs])
    },
    [password, setIsFirstTimeUser, setWallets, setAccounts]
  )

  const updateWallet = React.useCallback(
    async (id: string, wallet: UpdateWalletParams) => {
      const result = await sendMsgToChromeExt({
        event: 'updateWallet',
        data: {
          wallet,
          id,
          password,
        },
      })
      setWallets((ws) => ws.map((w) => (w.id === id ? result.wallet : w)))
    },
    [password, setWallets]
  )

  const deleteWallet = React.useCallback(
    async (id: string) => {
      await sendMsgToChromeExt({
        event: 'deleteWallet',
        data: {
          id,
          password,
        },
      })
      await Promise.all(
        accounts
          .filter((a) => a.walletId === id)
          .map((a) =>
            sendMsgToChromeExt({
              event: 'deleteAccount',
              data: {
                address: a.address,
                password,
              },
            })
          )
      )
      setWallets((ws) => {
        const newWallets = ws.filter((w) => w.id !== id)
        if (!newWallets.length) {
          setIsFirstTimeUser(true)
        }
        return newWallets
      })
      setAccounts((acs) => acs.filter((a) => a.walletId !== id))
    },
    [password, accounts, setWallets, setIsFirstTimeUser, setAccounts]
  )

  const addAccount = React.useCallback(
    async (account: CreateAccountParams, securityPassword: string) => {
      const result = await sendMsgToChromeExt({
        event: 'addAccount',
        data: {
          account,
          securityPassword,
          password,
        },
      })
      setAccounts((acs) => [result.account, ...acs])
    },
    [password, setAccounts]
  )

  const updateAccount = React.useCallback(
    async (address: string, account: UpdateAccountParams) => {
      const result = await sendMsgToChromeExt({
        event: 'updateAccount',
        data: {
          account,
          address,
          password,
        },
      })
      setAccounts((acs) => acs.map((a) => (a.address === address ? result.account : a)))
    },
    [password, setAccounts]
  )

  const deleteAccount = React.useCallback(
    async (address: string) => {
      await sendMsgToChromeExt({
        event: 'deleteAccount',
        data: {
          address,
          password,
        },
      })
      setAccounts((acs) => acs.filter((a) => a.address !== address))
    },
    [password, setAccounts]
  )

  // Can be used to verify security password
  const viewMnemonicPhrase = React.useCallback(
    async (id: string, securityPassword: string) => {
      const { mnemonic } = await sendMsgToChromeExt({
        event: 'viewMnemonicPhrase',
        data: {
          id,
          securityPassword,
          password,
        },
      })
      return mnemonic
    },
    [password]
  )

  const viewMnemonicPhraseBackup = React.useCallback(
    async (id: string, securityPassword: string, backupPassword: string) => {
      const { mnemonic } = await sendMsgToChromeExt({
        event: 'viewMnemonicPhraseBackup',
        data: {
          id,
          securityPassword,
          backupPassword,
          password,
        },
      })
      return mnemonic
    },
    [password]
  )

  React.useEffect(() => {
    checkIsFirstTimeUser()
  }, [])

  return (
    <WalletsContext.Provider
      value={{
        isFirstTimeUser,
        isUnlocked: !!wallets.length,
        wallets,
        accounts,
        password,
        unlockWallets,
        addWallet,
        updateWallet,
        deleteWallet,
        addAccount,
        updateAccount,
        deleteAccount,
        viewMnemonicPhraseBackup,
        viewMnemonicPhrase,
      }}
    >
      {children}
    </WalletsContext.Provider>
  )
}

const useWalletsContext = (): WalletsState => React.useContext(WalletsContext)

export { WalletsProvider, useWalletsContext }
