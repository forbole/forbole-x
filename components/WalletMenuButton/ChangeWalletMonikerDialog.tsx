import {
  Box,
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  IconButton,
  TextField,
  Typography,
} from '@material-ui/core'
import useTranslation from 'next-translate/useTranslation'
import React from 'react'
import CloseIcon from '../../assets/images/icons/icon_cross.svg'
import useStyles from './styles'
import { useWalletsContext } from '../../contexts/WalletsContext'
import useIconProps from '../../misc/useIconProps'

interface ChangeWalletMonikerDialogProps {
  walletId: string
  open: boolean
  onClose(): void
}

const ChangeWalletMonikerDialog: React.FC<ChangeWalletMonikerDialogProps> = ({
  walletId,
  open,
  onClose,
}) => {
  const { t } = useTranslation('common')
  const classes = useStyles()
  const iconProps = useIconProps()
  const [name, setName] = React.useState('')
  const { updateWallet } = useWalletsContext()

  const onButtonClick = React.useCallback(async () => {
    try {
      await updateWallet(walletId, { name })
      onClose()
    } catch (err) {
      console.log(err)
    }
  }, [name, updateWallet, walletId])

  React.useEffect(() => {
    if (open) {
      setName('')
    }
  }, [open])

  return (
    <Dialog fullWidth open={open} onClose={onClose}>
      <IconButton className={classes.closeButton} onClick={onClose}>
        <CloseIcon {...iconProps} />
      </IconButton>
      <DialogTitle>{t('change wallet moniker')}</DialogTitle>
      <DialogContent>
        <Box mb={18}>
          <Typography gutterBottom>{t('wallet moniker')}</Typography>
          <TextField
            fullWidth
            variant="filled"
            InputProps={{
              disableUnderline: true,
            }}
            placeholder={t('wallet moniker')}
            value={name}
            onChange={(e) => setName(e.target.value)}
          />
        </Box>
      </DialogContent>
      <DialogActions>
        <Button
          className={classes.dialogButton}
          variant="contained"
          color="primary"
          onClick={onButtonClick}
        >
          {t('next')}
        </Button>
      </DialogActions>
    </Dialog>
  )
}

export default ChangeWalletMonikerDialog
