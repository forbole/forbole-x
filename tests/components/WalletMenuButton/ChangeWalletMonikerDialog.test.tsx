import React from 'react'
import renderer from 'react-test-renderer'
import { useWalletsContext } from '../../../contexts/WalletsContext'
import ChangeWalletMonikerDialog from '../../../components/WalletMenuButton/ChangeWalletMoniker'

const onClose = jest.fn()
const updateWallet = jest.fn()

jest.mock('@material-ui/core/Dialog', () => (props) => <div id="dialog" {...props} />)
jest.mock('../../../contexts/WalletsContext', () => ({
  useWalletsContext: jest.fn(),
}))
;(useWalletsContext as jest.Mock).mockReturnValue({ updateWallet })

describe('component: ChangeWalletMonikerDialog', () => {
  it('renders opened state correctly', () => {
    const component = renderer.create(
      <ChangeWalletMonikerDialog walletId="123" onClose={onClose} />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  // it('renders closed state correctly', () => {
  //   const component = renderer.create(
  //     <ChangeWalletMonikerDialog walletId="123" open={false} onClose={onClose} />
  //   )
  //   const tree = component.toJSON()
  //   expect(tree).toMatchSnapshot()
  // })
  // it('calls onClose on close button click', () => {
  //   const component = renderer.create(
  //     <ChangeWalletMonikerDialog walletId="123" open onClose={onClose} />
  //   )
  //   renderer.act(() => {
  //     component.root.findAllByType('button')[0].props.onClick({ currentTarget: 'anchor' })
  //   })
  //   expect(onClose).toBeCalled()
  // })
  // it('calls updateWallet on next button click', async () => {
  //   const component = renderer.create(
  //     <ChangeWalletMonikerDialog walletId="123" open onClose={onClose} />
  //   )
  //   renderer.act(() => undefined)
  //   renderer.act(() => {
  //     component.root.findByType('input').props.onChange({ target: { value: 'new name' } })
  //   })
  //   await renderer.act(async () => {
  //     await component.root.findAllByType('button')[1].props.onClick()
  //   })
  //   expect(updateWallet).toBeCalledWith('123', { name: 'new name' })
  //   expect(onClose).toBeCalled()
  // })
  // it('handles error on next button click', async () => {
  //   updateWallet.mockRejectedValueOnce(new Error('error'))
  //   const component = renderer.create(
  //     <ChangeWalletMonikerDialog walletId="123" open onClose={onClose} />
  //   )
  //   renderer.act(() => undefined)
  //   renderer.act(() => {
  //     component.root.findByType('input').props.onChange({ target: { value: 'new name' } })
  //   })
  //   await renderer.act(async () => {
  //     await component.root.findAllByType('button')[1].props.onClick()
  //   })
  //   expect(updateWallet).toBeCalledWith('123', { name: 'new name' })
  //   expect(onClose).toBeCalledTimes(0)
  // })
})

afterEach(() => {
  jest.clearAllMocks()
})
