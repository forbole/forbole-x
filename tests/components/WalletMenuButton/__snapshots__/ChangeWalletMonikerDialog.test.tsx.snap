// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: ChangeWalletMonikerDialog renders closed state correctly 1`] = `
<div
  fullWidth={true}
  id="dialog"
  onClose={[MockFunction]}
  open={false}
>
  <button
    className="MuiButtonBase-root MuiIconButton-root HookGlobalStyles-closeButton-4"
    disabled={false}
    onBlur={[Function]}
    onClick={[MockFunction]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        data-file-name="SvgIcon_cross"
        fill="#e0e0e0"
        height={16}
        width={16}
      />
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>
  <div
    className="MuiDialogTitle-root"
  >
    <h2
      className="MuiTypography-root MuiTypography-h6"
    >
      common:change wallet moniker
    </h2>
  </div>
  <div
    className="MuiDialogContent-root"
  >
    <div
      className="MuiBox-root MuiBox-root-11"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
      >
        common:wallet moniker
      </p>
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <div
          className="MuiInputBase-root MuiFilledInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiFilledInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="common:wallet moniker"
            required={false}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiDialogActions-root MuiDialogActions-spacing"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained HookGlobalStyles-dialogButton-3 MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        common:next
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`component: ChangeWalletMonikerDialog renders opened state correctly 1`] = `
<div
  fullWidth={true}
  id="dialog"
  onClose={[MockFunction]}
  open={true}
>
  <button
    className="MuiButtonBase-root MuiIconButton-root HookGlobalStyles-closeButton-4"
    disabled={false}
    onBlur={[Function]}
    onClick={[MockFunction]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        data-file-name="SvgIcon_cross"
        fill="#e0e0e0"
        height={16}
        width={16}
      />
    </span>
  </button>
  <div
    className="MuiDialogTitle-root"
  >
    <h2
      className="MuiTypography-root MuiTypography-h6"
    >
      common:change wallet moniker
    </h2>
  </div>
  <div
    className="MuiDialogContent-root"
  >
    <div
      className="MuiBox-root MuiBox-root-10"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
      >
        common:wallet moniker
      </p>
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <div
          className="MuiInputBase-root MuiFilledInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiFilledInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="common:wallet moniker"
            required={false}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiDialogActions-root MuiDialogActions-spacing"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained HookGlobalStyles-dialogButton-3 MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        common:next
      </span>
    </button>
  </div>
</div>
`;
